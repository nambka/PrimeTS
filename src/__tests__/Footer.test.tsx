import React from 'react';
import Footer from '../components/Footer';
import { shallow } from 'enzyme';

describe('Footer', () => {
  const logoImageAlt = 'Landing Page Logo';
  const props = {
    theme: '',
    toggleTheme: jest.fn()
  }
  beforeEach(() => {
    // const wrapper = mount(<Footer />);
  });

  afterEach(() => {
    jest.resetModules();
    // wrapper.unmount();
  });

  it('renders nav items without errors', () => {
    const wrapper = shallow(<Footer {...props}/>);
    expect(wrapper.find('div.container').length).toEqual(1);
    expect(wrapper.find('span.h5').length).toEqual(1);
    expect(wrapper.find('#btnDarkModeFooter').length).toEqual(1);
    expect(wrapper).toMatchSnapshot();
  });

  // Ref: https://stackoverflow.com/questions/43747397/simulate-a-button-click-in-jest
  it('simulates DarkMode button click event', () => {
    const mockDarkMode = jest.fn();
    const wrapper = shallow(<Footer {...props} toggleTheme={mockDarkMode}/>);
    wrapper.find('button').simulate('click');
    expect(mockDarkMode).toHaveBeenCalled();
  });

  it('should change DarkMode icon to Sun', () => {
    const mockDarkMode = jest.fn();
    const theme = 'light';
    const useStateSpy = jest.spyOn(React, 'useState');
    useStateSpy.mockImplementation((init) => [init, mockDarkMode]);
    const wrapper = shallow(<Footer {...props} theme={theme} toggleTheme={mockDarkMode}/>);

    wrapper.find('button').simulate('click');
    // console.log(wrapper.debug());
    expect(mockDarkMode).toHaveBeenCalled();
    expect(wrapper.find('i').last().prop('className')).toEqual('fa fa-sun-o');
  });

  it('should change DarkMode icon to Circle', () => {
    const mockDarkMode = jest.fn();
    const theme = 'dark';
    const useStateSpy = jest.spyOn(React, 'useState');
    useStateSpy.mockImplementation((init) => [init, mockDarkMode]);
    const wrapper = shallow(<Footer theme={theme} toggleTheme={mockDarkMode}/>);

    wrapper.find('button').simulate('click');
    expect(mockDarkMode).toHaveBeenCalled();
    expect(wrapper.find('i').last().prop('className')).toEqual('fa fa-circle-o');
  });

  it('should change DarkMode icon to Moon', () => {
    const mockDarkMode = jest.fn();
    const theme = 'moon';
    const useStateSpy = jest.spyOn(React, 'useState');
    useStateSpy.mockImplementation((init) => [init, mockDarkMode]);
    const wrapper = shallow(<Footer theme={theme} toggleTheme={mockDarkMode}/>);

    wrapper.find('button').simulate('click');
    expect(mockDarkMode).toHaveBeenCalled();
    expect(wrapper.find('i').last().prop('className')).toEqual('fa fa-moon-o');
  });

  // https://blog.logrocket.com/testing-state-changes-in-react-functional-components/
  it('should update state on DarkMode button clicked', () => {
    const mockDarkMode = jest.fn();
    const handleClick = jest.spyOn(React, 'useState');
    const wrapper = shallow(<Footer toggleTheme={mockDarkMode}/>);
    handleClick.mockImplementation(theme => [theme, mockDarkMode]);
 
    wrapper.find("#btnDarkModeFooter").simulate("click");
    expect(mockDarkMode).toBeTruthy();
  });

  it('renders img without errors', () => {
    const wrapper = shallow(<Footer />);
    expect(wrapper.find('img').length).toEqual(1);
    expect(wrapper.find('img').prop('alt')).toEqual(logoImageAlt);
    // expect(wrapper.find('img').getElement(0).props.src).toEqual(logoImage);
    // expect(wrapper.find('img').prop('src')).toEqual(logoImage);
    // expect(wrapper.find('img').props().src).toEqual(logoImage);
  });

})