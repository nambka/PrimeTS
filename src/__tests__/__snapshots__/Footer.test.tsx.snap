// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Footer renders nav items without errors 1`] = `
<footer
  className="gradient-half-primary-body-v1 py-4"
  style={
    Object {
      "height": "calc(100% - 60px)",
    }
  }
>
  <div
    className="container"
  >
    <div
      className="row justify-content-md-between font-size-1 py-3"
    >
      <div
        className="col-lg-3 mb-4 mb-lg-0"
      >
        <div
          className="d-flex h-100 flex-column"
        >
          <div
            className="d-flex align-items-center mb-4"
          >
            <img
              alt="Landing Page Logo"
              src={null}
              width="32"
            />
            <span
              className="h5 mb-0 ml-3"
            >
              Powered by Nambk
            </span>
          </div>
          <p
            className="font-size-0-75"
          >
            A brand new journey to the beautiful angles of this world. Neque porro quisquam est qui dolorem ipsum quia dolor sit amet, consectetur, adipisci velit...
          </p>
          <div
            className="d-flex align-items-center"
          >
            <a
              className="btn btn-xss btn-soft-light mr-2"
              href="https://github.com/nambka/PrimeTS"
            >
              <i
                className="fa fa-cogs mr-1"
              />
              Preferences
            </a>
            <button
              aria-label="Dark/Light Mode"
              className="btn btn-sm btn-icon btn-soft-light"
              data-original-title=""
              data-title="Day/Night Mode"
              data-toggle="tooltip"
              id="btnDarkModeFooter"
              onClick={[Function]}
            >
              <i
                className="fa fa-moon-o"
              />
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</footer>
`;
